harvester/harvester
# Add upgrade webhook

**Problem:**
We need to add some webhook validation checks to forbid users from performing upgrades in some cases.
This gives the user a smooth upgrade experience and ensures data reliability.

**Related Issue:**
https://github.com/harvester/harvester/issues/2833

**Test plan:**

Case 1: A degraded volume in 3-node cluster can't upgrade.
1. Create a harvester cluster (node >= 3).
2. Create a VM with a big enough volume. Make sure the robustness of the volume is `degraded`.
3. Create a `version` CR.

```
cat <<EOF | kubectl apply -f -
apiVersion: harvesterhci.io/v1beta1
kind: Version
metadata:
  name: v1.1.0
  namespace: harvester-system
spec:
  isoURL: "http://192.168.122.1:8000/harvester-c76dfbd-dirty-amd64.iso"
EOF
```

4. Click the "Upgrade" button and choose "v1.1.0". We can't do upgrade.

![Screen Shot 2022-09-28 at 3 35 24 PM](https://user-images.githubusercontent.com/4927361/192739778-e5279d46-f3e5-4e46-8dec-fd5c8e0e6896.png)

Case 2: A `provisioning` cluster can't upgrade.
1. Follow case 1. Clean up the VM and degraded volumes.
2. Click the "Upgrade" button and choose "v1.1.0".
3. Update `spec.rkeConfig.provisionGeneration` in `fleet-local/local` clusters.provisioning.cattle.io.
```
spec:
  rkeConfig:
    provisionGeneration: 1
```
4. When the phase of `fleet-local/local` clusters.cluster.x-k8s.io is `Provisioning`, delete the upgrade. The upgrade can't be deleted.

Case 3: An upgrade with the `harvesterhci.io/skipWebhook: "true"` annotation can be deleted.
1. Follow case 2.
2. Update the upgrade with annotation `harvesterhci.io/skipWebhook: "true"`.
3. If the phase of `fleet-local/local` clusters.cluster.x-k8s.io is `Provisioned`, you can update `spec.rkeConfig.provisionGeneration` in `fleet-local/local` clusters.provisioning.cattle.io to trigger it becomes `Provisioning`.
4. Delete the upgrade. This time upgrade CR can be deleted.
